spec.html

```diff
@@ -20670,26 +21469,41 @@ <h1>Runtime Semantics: ClassDefinitionEvaluation</h1>
+        1. Set the running execution context's PrivateEnvironment to _classPrivateEnvironment_.
         1. Let _constructorInfo_ be ! DefineMethod of _constructor_ with arguments _proto_ and _constructorParent_.
         1. Let _F_ be _constructorInfo_.[[Closure]].
         1. Perform SetFunctionName(_F_, _className_).
         1. Perform MakeConstructor(_F_, *false*, _proto_).
         1. If |ClassHeritage_opt| is present, set _F_.[[ConstructorKind]] to ~derived~.
         1. Perform MakeClassConstructor(_F_).
         1. Perform CreateMethodProperty(_proto_, *"constructor"*, _F_).
-        1. If |ClassBody_opt| is not present, let _methods_ be a new empty List.
-        1. Else, let _methods_ be NonConstructorMethodDefinitions of |ClassBody|.
-        1. For each |ClassElement| _m_ of _methods_, do
-          1. If IsStatic of _m_ is *false*, then
-            1. Let _status_ be PropertyDefinitionEvaluation of _m_ with arguments _proto_ and *false*.
+        1. If |ClassBody_opt| is not present, let _elements_ be a new empty List.
+        1. Else, let _elements_ be NonConstructorMethodDefinitions of |ClassBody|.
+        1. Let _instanceFields_ be a new empty List.
+        1. Let _staticFields_ be a new empty List.
+        1. For each |ClassElement| _e_ in order from _elements_, do
+          1. If IsStatic of _e_ is *false*, then
+            1. Let _field_ be ClassElementEvaluation of _e_ with arguments _proto_ and *false*.
           1. Else,
-            1. Let _status_ be PropertyDefinitionEvaluation of _m_ with arguments _F_ and *false*.
-          1. If _status_ is an abrupt completion, then
+            1. Let _field_ be ClassElementEvaluation of _e_ with arguments _F_ and *false*.
+          1. If _field_ is an abrupt completion, then
             1. Set the running execution context's LexicalEnvironment to _env_.
-            1. Return Completion(_status_).
+            1. Set the running execution context's PrivateEnvironment to _outerPrivateEnvironment_.
+            1. Return Completion(_field_).
+          1. If _field_ is a ClassFieldDefinition Record, then
+            1. If IsStatic of _e_ is *false*, append _field_ to _instanceFields_.
+            1. Else, append _field_ to _staticFields_.
         1. Set the running execution context's LexicalEnvironment to _env_.
         1. If _classBinding_ is not *undefined*, then
           1. Perform _classScope_.InitializeBinding(_classBinding_, _F_).
+        1. Set _F_.[[Fields]] to _instanceFields_.
+        1. If PrivateBoundIdentifiers of |ClassBody| contains a Private Name _P_ such that _P_.[[Kind]] is either ~method~ or ~accessor~ and _P_.[[Brand]] is _proto_, then
+          1. Set _F_.[[PrivateBrand]] to _proto_.
+        1. If PrivateBoundIdentifiers of |ClassBody| contains a Private Name _P_ such that _P_.[[Kind]] is either ~method~ or ~accessor~ and _P_.[[Brand]] is _F_, then
+          1. Perform ? PrivateBrandAdd(_F_, _F_).
```

Comment:
The `ReturnIfAbrupt` expansion of `?` short-circuits the restoration of the running execution context's PrivateEnvironment at the end of this algorithm. As written, the private environment will leak.